File: FunctionalConcurrencyNET/Chapter.14/StockTicker.Server/Core/TradingAgent.fs
[<AutoOpenAttribute>]
module TradingAgent

open System
open Microsoft.AspNet.SignalR.Hubs
open StockTicker.Core
open StockTicker.Server

// specialized agent for each user
// keep track orders and status of portfolio
type TradingAgent(connId : string, initialAmount : float, caller:IHubCallerConnectionContext<IStockTickerHubClient>) =  // #A

    let agent =
        new Agent<Trading>(fun inbox ->
            // single thread safe no sharing
            let rec loop cash (portfolio : Portfolio) (buyOrders : Treads) (sellOrders : Treads) =
                async {  // #B
                    let! msg = inbox.Receive()
                    match msg with
                    | Kill(reply) -> reply.Reply()   // #C
                    | Error(exn) -> raise exn    // #C

                    | Trading.Buy(symbol, trading) ->    // #D
                        let items = setOrder buyOrders symbol trading 
                        let order = createOrder symbol trading TradingType.Buy
                        caller.Client(connId).UpdateOrderBuy(order)
                        return! loop cash portfolio items sellOrders

                    | Trading.Sell(symbol, trading) ->    // #D
                        let items = setOrder sellOrders symbol trading
                        let order = createOrder symbol trading TradingType.Sell
                        caller.Client(connId).UpdateOrderSell(order)
                        return! loop cash portfolio buyOrders items

                    | Trading.UpdateStock(stock) ->     // #E
                        caller.Client(connId).UpdateStockPrice stock

                        let cash, portfolio, sellOrders = updatePortfolio cash stock portfolio sellOrders TradingType.Sell
                        let cash, portfolio, buyOrders = updatePortfolio cash stock portfolio buyOrders TradingType.Buy
                          
                        let asset = getUpdatedAsset portfolio sellOrders buyOrders cash   // #F
                        caller.Client(connId).UpdateAsset(asset)    // #G

                        return! loop cash portfolio buyOrders sellOrders
            }
            loop initialAmount (Portfolio(HashIdentity.Structural)) (Treads(HashIdentity.Structural))
                (Treads(HashIdentity.Structural)))

    member x.Agent = agent

    interface IObserver<Trading> with   // #H
        member x.OnNext(msg) = agent.Post(msg:Trading)      // #I
        member x.OnError(exn) = agent.Post(Error exn)       // #C
        member x.OnCompleted() = agent.PostAndReply(Kill)   // #C

File: FunctionalConcurrencyNET/Chapter.14/StockTicker.Server.Cs/Core/TradingAgent.cs
using Microsoft.AspNet.SignalR.Hubs;
using ReactiveAgent.Agents;
using ReactiveAgent.Agents.Dataflow;
using StockTicker.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Web;
using static StockTicker.Core.Models;

namespace StockTicker.Server.Cs.Core
{
    public class TradingAgent : IObserver<Trading>
    {
        class State
        {
            public double Cash { get; set; }
            public Dictionary<string, TradingDetails> Portfolio { get; set; }
            public Dictionary<string, List<TradingDetails>> BuyOrders { get; set; }
            public Dictionary<string, List<TradingDetails>> SellOrders { get; set; }
        }

        public TradingAgent (string connId, double initialAmount, IHubCallerConnectionContext<IStockTickerHubClient> caller)
        {
            var initialState = new State
            {
                Cash = initialAmount,
                Portfolio = new Dictionary<string, TradingDetails>(),
                BuyOrders = new Dictionary<string, List<TradingDetails>>(),
                SellOrders = new Dictionary<string, List<TradingDetails>>()
            };

            agent = Agent.Start<State, Trading>(initialState,
                (state, message)=>
                {
                    if (message is Trading.Kill)
                    {
                        throw new NotImplementedException("Kill messages should not be send to Dataflow based Agents");
                    }
                    else if (message is Trading.Error msgError)
                    {
                        throw msgError.Item;
                    }
                    else if (message is Trading.Buy msgBuy)
                    {
                        var items = HelperFunctions.setOrder(state.BuyOrders, msgBuy.symbol, msgBuy.Item2);
                        var order = HelperFunctions.createOrder(msgBuy.symbol, msgBuy.Item2, TradingType.Buy);
                        caller.Client(connId).UpdateOrderBuy(order);
                        state.BuyOrders = items;
                    }
                    else if (message is Trading.Sell msgSell)
                    {
                        var items = HelperFunctions.setOrder(state.SellOrders, msgSell.symbol, msgSell.Item2);
                        var order = HelperFunctions.createOrder(msgSell.symbol, msgSell.Item2, TradingType.Sell);
                        caller.Client(connId).UpdateOrderSell(order);
                        state.SellOrders = items;
                    }
                    else if (message is Trading.UpdateStock msgStock)
                    {
                        caller.Client(connId).UpdateStockPrice(msgStock.Item);

                        var x = HelperFunctions.updatePortfolio(state.Cash, msgStock.Item, state.Portfolio, state.SellOrders, TradingType.Sell);
                        state.Cash = x.Item1;
                        state.Portfolio = x.Item2;
                        state.SellOrders = x.Item3;

                        var y = HelperFunctions.updatePortfolio(state.Cash, msgStock.Item, state.Portfolio, state.BuyOrders, TradingType.Buy);
                        state.Cash = y.Item1;
                        state.Portfolio = y.Item2;
                        state.BuyOrders = y.Item3;

                        var asset = HelperFunctions.getUpdatedAsset(state.Portfolio, state.SellOrders, state.BuyOrders, state.Cash);
                        caller.Client(connId).UpdateAsset(asset);
                    }
                    return state;
                });
        }

        private readonly IAgent<Trading> agent;

        public IAgent<Trading> GetAgent => agent;

        public void OnNext(Trading value)
            => agent.Post(value);

        public void OnError(Exception error)
            => agent.Post(Trading.NewError(error));

        public void OnCompleted()
        {
            if (agent is IDisposable killer)
                killer.Dispose();
            else
                throw new Exception("Wrong agent is used (that cannot be killed)");
        }
    }
}